### ISSUE TITLE
Replace Regex crashes when using closing braces 

### ISSUE THREAD

**Issue by Timboo89** [hashicorp/terraform#18672]
Hi there, 

I want to generate environment variables of ECS task definitions with terraform maps, unfortunately this is not so easy in TF. So my plan is the following.
- Create a flat list, that contains each key followed by its value (flat_envvars -> works fine)
- Create a flat list, that just contains ["name", "value"] as often, as there are key value pairs (not_expected_keys -> is not working because this [1] is added after each element, removing (workaround_keys) this does not work because I cannot use a closing brace in a regex. [2] What I tried
- zip these lists in a map

[1]
```
%!(EXTRA string=key1,
     string=val1)
```
[2]
```
/%!.*)/
/%!.*\)/
/%!.*\\)/
/%!.*\\\)/
```

### Terraform Version
```
v0.11.7
```

### Terraform Configuration Files
```hcl
locals {
  envvars = {
    "key1" = "val1",
    "key2" = "val2"
  },
}

output "flat_envvars" {
  value = "${split(",", join(",", formatlist("%s,%s", keys(local.envvars), values(local.envvars))))}"
}

output "not_expected_keys" {
  value = "${formatlist("name,value", keys(local.envvars), values(local.envvars))}"
}

output "workaround_keys" {
  value = "${split(",", replace(join(",", formatlist("name,value", keys(local.envvars), values(local.envvars))), "/%!.*)/", ","))}"
}
```

### Expected Behavior
Just replace the damn string ;-)

### Actual Behavior
`replace: error parsing regexp: unexpected ): ``%!.*)`` in:`

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

**Comment by Timboo89** [comment]
My current workaround is this ugly stuff.
`"${split(",", replace(replace(join(",", formatlist("name,value", keys(local.envvars))), local.search_trailing_comma, ","), local.search_no_trailing_comma, " "))}"`

**Comment by apparentlymart** [comment]
Hi @Timboo89! Sorry for the slow response here.

I just tried your configuration as given in Terraform v0.12.0-alpha4 and got a more verbose version of the same error, just as a result of the general error message improvements from Terraform v0.12:

```
Error: Error in function call

  on replace-regexp-error.tf line 17, in output "workaround_keys":
  17:   value = "${split(",", replace(join(",", formatlist("name,value", keys(local.envvars), values(local.envvars))), "/%!.*)/", ","))}"
    |----------------
    | local.envvars is object with 2 attributes

Call to function "replace" failed: error parsing regexp: unexpected ): `%!.*)`.
```

However, on closer inspection I saw that this error is expected, because there is indeed a literal closing parenthesis in the pattern there. Since parentheses are part of the regular expression syntax, a literal parenthesis must be escaped with a backslash. Since backslashes are also used for escapes in quoted strings, we need to escape it twice to get the backslash through to the regular expression parser:

```hcl
"/%!.*\\)/"
```

With that said, Terraform v0.12 also offers a new way to write the projection from a simple map into a list of objects with name/value attributes:

```hcl
output "like_ecs_wants" {
  value = [for k, v in local.envvars : {name = k, value = v}]
}
```

```
like_ecs_wants = [
  {
    "name" = "key1"
    "value" = "val1"
  },
  {
    "name" = "key2"
    "value" = "val2"
  },
]
```

This avoids the need for all of those nested function calls. This feature is in master ready to be included in the forthcoming v0.12.0 release, so I'm going to close this out. Sorry we didn't notice the errant parenthesis there earlier!
